azure-iot-device/azure/iot/device/provisioning/pipeline/pipeline_stages_provisioning_mqtt.py

```diff
@@ -0,0 +1,121 @@
+        super(ProvisioningMQTTConverter, self).__init__()
+        self.action_to_topic = {}
+
+    def _run_op(self, op):
+
+        if isinstance(op, pipeline_ops_provisioning.SetSymmetricKeySecurityClientArgs):
+            # get security client args from above, save some, use some to build topic names,
+            # always pass it down because MQTT Provider stage will also want to receive these args.
+
+            client_id = op.registration_id
+            client_version = urllib.parse.quote_plus(constant.USER_AGENT)
+            username = "{id_scope}/registrations/{registration_id}/api-version={api_version}&ClientVersion={client_version}".format(
+                id_scope=op.id_scope,
+                registration_id=op.registration_id,
+                api_version=constant.API_VERSION,
+                client_version=client_version,
+            )
+
+            hostname = op.provisioning_host
+
+            self.continue_with_different_op(
+                original_op=op,
+                new_op=pipeline_ops_mqtt.SetConnectionArgs(
+                    client_id=client_id, hostname=hostname, username=username
+                ),
+            )
+
+        elif isinstance(op, pipeline_ops_provisioning.SendRegistrationRequest):
+            # Convert Sending the request into Mqtt Publish operations
+            topic = mqtt_topic.get_topic_for_register(op.rid)
+            self.continue_with_different_op(
+                original_op=op, new_op=pipeline_ops_mqtt.Publish(topic=topic, payload=op.request)
+            )
+
+        elif isinstance(op, pipeline_ops_provisioning.SendQueryRequest):
+            # Convert Sending the request into Mqtt Publish operations
+            topic = mqtt_topic.get_topic_for_query(op.rid, op.operation_id)
+            self.continue_with_different_op(
+                original_op=op, new_op=pipeline_ops_mqtt.Publish(topic=topic, payload=op.request)
+            )
+
+        # TODO : In the iothub case the enable and disable operations are defined in base.
+        # TODO: I felt this is very specific to DPS so this is defined in provisioning
+        elif isinstance(op, pipeline_ops_base.EnableFeature):
+            # Enabling for register gets translated into an Mqtt subscribe operation
+            topic = mqtt_topic.get_topic_for_subscribe()
+            self.continue_with_different_op(
+                original_op=op, new_op=pipeline_ops_mqtt.Subscribe(topic=topic)
+            )
+
+        elif isinstance(op, pipeline_ops_base.DisableFeature):
+            # Disabling a register response gets turned into an Mqtt unsubscribe operation
+            topic = mqtt_topic.get_topic_for_subscribe()
+            self.continue_with_different_op(
+                original_op=op, new_op=pipeline_ops_mqtt.Unsubscribe(topic=topic)
+            )
+
+        else:
+            # All other operations get passed down
+            self.continue_op(op)
+
+    def _handle_pipeline_event(self, event):
```

Comment:
This method is the best place to start data encapsulation (if possible)

Reply:
its actually not possible here. It is not advisable to create a DPS specific response and send. The request response provider may become a generic pipeline much like what the Twin has now. In that model a very generic response needs to be sent upwards.

---
In reply to: [287547898](https://github.com/Azure/azure-iot-sdk-python-preview/pull/97#discussion_r287547898) [](ancestors = 287547898)